### PR TITLE
Jupyter notebooks for three tutorials

### PR DESCRIPTION
**Pull request by tomdemeyere**
@JacekDziedzic 

Some progress for the jupyter notebooks

The main thing to look at here is the organization of the pseudopotentials and files, are we happy with the current hierarchy?

### CODE REVIEW

**Comment by JacekDziedzic** [comment]
Hi @tomdemeyere. Thanks for these changes. @cksdaddy is busy with other work and tasked me with looking at these jupyterified tutorials. Which might be a good thing, since I am 100% jupyter-naive. I had other responsibilities earlier, so sorry for the delay, but I'm on it now.

I was told that Jupyter is best used from VSCode, so I installed that on my Windows box. @BCAyers2000 and @Nekkrad helped me set up python and Jupyter in VSCode on this box, and that worked. Trying to run the first cell in the first tutorial told me that the ASE module is missing, so I did `pip install ase` from the command line and this seems to have helped. The second cell, I imagine, sets up the pseudo path and the command to launch ONETEP. This, I believe, is where my Windows adventure ends, since I don't have a ONETEP executable for Windows, and I know building one will be a nightmare.

I then switched to Linux. For now I omitted VSCode and thought I'd go the easiest route, via a browser. I did `sudo apt install jupyter-core` and `sudo apt install jupyter-notebook` and that, to my understanding, lets me open your tutorials in a browser. So far, so good. My next step was `pip install ase`, since it was absent, just like on Windows.

The 1st cell then ran fine. My attempt to run the 2nd cell finished with:
ase_error

Can you help me with that? I have no idea why `pseudo_path` would be unrecognized by, presumably, the ONETEP calculator.

Also, would it be possible to make the launching script use the ONETEP launcher rather than directly starting the executable? I don't want the users to be confused -- we teach them to use the launcher from the very beginning. Also, can the path be made into a variable defined earlier, with some comment to adjust `/local/disk2/tom_stuff_please_do_not_delete/onetep` to something suitable?

**Comment by tomdemeyere** [comment]
Hello Jacek, I hope you are doing well,

The first problem is due to the fact that `pseudo_path` was introduced only recently so you will have to install the git version of ase, in your case (a new release should come before the master class)

``` bash
pip install --force-reinstall --no-deps git+https://gitlab.com/ase/ase.git@master
```

**A little bit more context about launching ONETEP with ASE if you have time/willingness:**

The problem with these tutorials in a Jupyter notebook (and an area where you might help) is that ASE make use of the python module `subprocess` and especially the `check_call` [command](https://docs.python.org/3/library/subprocess.html#subprocess.check_call)

Using this function there are two possibilities to run a program, either python start an intermediate shell (shell=True) and the program is called there. The other possibility (shell=False) will tell Python to start the program itself, via an OS call.

In practice, the first option is never used: there are too many security issues, also it does not work well with Windows, so ASE uses `shell=False`, hardcoded.

Why am I telling you all that? Because the problem with the latter approach is that you can't use anything normally defined during shell init (bash_profile, bashrc or else) hence, you can say bye bye to any defined environment variables defined or even the module command for example.

To avoid this issue, it is possible to send a list of env variables to the `check_call` command which is what ASE does. It sends the current env variables as context. So when you setup a slurm script for example, you load all the modules, and invoke python, everything will work fine.

But in the realm of Jupyter Notebooks this solution does not work because the jupyter notebook itself has its own weird shell and nothing is defined there, so you send empty context to `check_call`. The only solution I found to this problem is to instead call a script `launch_onetep.sh` which in the case of RH9 load the necessary module and launch the script that defines all the env variable, ONETEP is then launched within that script, and everything works.

The other solution would be to somewhat hardcode all relevant environment variables and put them in the Jupyter notebook, but since everyone will have different config that's not possible.

So in conclusion, yes I would like to use the `onetep_launcher` but I don't think it has the ability to load modules and script?

**Comment by JacekDziedzic** [comment]
Thanks for the rapid response, @tomdemeyere. I tried the new `pip install` command you provided, but it didn't help, i.e. I'm getting the same error. I can't even tell if the `pip install` worked. It gave me this:

```
Defaulting to user installation because normal site-packages is not writeable
Collecting git+https://gitlab.com/ase/ase.git@master
  Cloning https://gitlab.com/ase/ase.git (to revision master) to /tmp/pip-req-build-zuim3dgk
  Running command git clone --filter=blob:none --quiet https://gitlab.com/ase/ase.git /tmp/pip-req-build-zuim3dgk
  Resolved https://gitlab.com/ase/ase.git to commit b132b03066352b76e98ca4b70e558b6d14a84b02
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
Building wheels for collected packages: UNKNOWN
  Building wheel for UNKNOWN (pyproject.toml) ... done
  Created wheel for UNKNOWN: filename=UNKNOWN-0.0.0-py3-none-any.whl size=18091 sha256=b8ddcbd10809ed8ae87bca70c543dede0fd8a758ac614adfa761854b31fb652f
  Stored in directory: /tmp/pip-ephem-wheel-cache-j_xbvtao/wheels/01/fc/60/e6134958a4ba6921161034c849be512a068712df0013719724
Successfully built UNKNOWN
Installing collected packages: UNKNOWN
Successfully installed UNKNOWN-0.0.0
```

The `UNKNOWN` part looks scary to me.

I think I understand your explanation about the shells. I am fine with the `.launch_onetep.sh` script, which, to my understanding, solves the problems you outlined. What I mean is, why does this script manually set `OMP_NUM_THREADS`, `ulimit -s` etc, rather than just calling `onetep_launcher` which does all this for you?

**Comment by tomdemeyere** [comment]
@JacekDziedzic It is indeed not normal, I hate to say it but: it does work on my machine. I suspect this is a python version related issue, check the results of both commands:

``` bash
pip --version
python --version
```
This should be >3.8 and both should match

**Comment by JacekDziedzic** [comment]
Thanks for bearing with me, @tomdemeyere. I get

```
jacek@Xubuntu22:~$ pip --version
pip 22.0.2 from /usr/lib/python3/dist-packages/pip (python 3.10)
jacek@Xubuntu22:~$ python --version
Python 3.10.12
```

**Comment by JacekDziedzic** [comment]
Hi again, @tomdemeyere. I tried on a fresh install of Ubuntu 24.04. This time the error message is more informative:
ase_error_2

Do I follow the advice they give?

**Comment by tomdemeyere** [comment]
Hi @JacekDziedzic, it's the system wide python nicely telling you "please don't touch me", I think you can bypass it by adding "--break-system-packages" but as the name suggest, you are at risk to break stuff.

One of the solution is to use venv (I never used them so I can't really help as much here), same for pipx but you can try it, it should work.

The solution that I almost always recommend is to say bye bye to system-wide python and install miniforge (https://github.com/conda-forge/miniforge?tab=readme-ov-file#download) you will have your things running in 2 minutes.

EDIT: btw, I am always available for a call, if you think this would be faster

**Comment by JacekDziedzic** [comment]
Hi @tomdemeyere,

I finally found more time to look at these. So far I went through tutorial 1, but the difficulties were mostly related to getting started, so I should be able to progress much faster now.

Brad and Weibo had to help me set this up. We used mambaforge to install jupyter, as my systemwide install seemingly had trouble choosing the correct ipykernel. We used ``pip install`` to get the correct ASE. Your ``--no-deps`` proved counterproductive -- it caused the install not to fetch ``numpy``. I am at the state where I can run the notebooks from firefox on my fresh linux install.

I have overall good impressions on tutorial 1. Here I will only list a few issues that I think should be corrected.

- I don't like that the ``.launch_onetep.sh`` script that is necessary to start ONETEP from ASE is hidden. It's a crucial part of the machinery and as such should be exposed. Suggest renaming it to something like ``ase_launch_onetep.sh``, to make it clear its purpose to interface with ASE.

- This script is not documented or commented at all, and it includes things very specific to your config. A few comments explaining the purpose of this thing, "load all required modules here", "set up the number of processes and threads here", etc. Number of processes and threads should be variables, defined at the top for better user experience. Using ``onetep_launcher`` instead of the bare binary would help make this clean.

- I think both paths in the profile are incorrect. I found that I had to add an extra "../" to the one you have, because we operate from a fresh directory ``t1_onetep_silane`` that gets created.

- A short preamble document would be useful that would explain the need to get the newer ASE. "Unexpected keyword argument: pseudo_path" is not an error message that lets users infer that. If not for your help, I would have never figured out I need to install a newer ASE.

I'll try to take a look, again, at the bulk silicon that does not converge. We can't have that. @cksdaddy , any suggestions? Maybe we can up the NGWF convergence threshold a little bit?

**Comment by JacekDziedzic** [comment]
Here's my suggestion for the script:

```
#!/bin/bash

############################################################
# This simple script is what ASE will use to start ONETEP. #
# You will need to adjust it to your environment.          #
############################################################

# Select the number of processes and threads.
# ONETEP will run on nprocesses * nthreads CPU cores.
nprocesses=2
nthreads=2

# Point this to your ONETEP executable 
# (normally found in your installation's ./bin directory).
onetep_exe=/home/user/onetep/bin/onetep.gfortran

# Point this to your onetep_launcher
# (normally found in your installation's ./utils directory).
onetep_launcher=/home/user/onetep/utils/onetep_launcher

# Load any necessary modules here, set up any necessary environment variables.
# e.g. for Intel you might have to
# module load oneapi
# source /local/software/intel/oneapi/2023.0/setvars.sh

# This runs ONETEP. On some systems you may need to replace 'mpirun'
# with a suitable alternative.
mpirun -np 2 $onetep_launcher -t $nthreads -e $onetep_exe $@
```

**Comment by JacekDziedzic** [comment]
The silane part creates a separate directory, which means the paths to the pseudos and the script need ``../../``. The silicon part operates in the current directory, which means the paths to the pseudos and the script only need ``../``. I think the silicon part should also be creating its own directory for consistency.

**Comment by JacekDziedzic** [comment]
I looked at the silicon example. No wonder it's not converging, **the structure is wrong**. The nearest-neighbour distance in bulk Si is about 2.35 \AA. The structure you have in the notebook has a nearest-neighbour distance of 4.40 \AA! Bohr vs angstroem?

**Comment by tomdemeyere** [comment]
@JacekDziedzic Well spotted, for some reason I told ASE that the lattice constant was >10 (I think that github copilot did it automatically and I didn't question it, AI is taking over) I set it up to something much more reasonable.

Thanks for all the suggestions, I attempted to apply them

**Comment by JacekDziedzic** [comment]
Hi @tomdemeyere -- thanks for the fixes and sorry I'm only looking at this in leaps and bounds. I've been busy with continuous integration all week.

I am now unable to run it at all due to a discrepancy between the pseudos the tutorial seems to be using:
```
Si_ONCV_PBE-1.2.upf
H_ONCV_PBE-1.2.upf
```
and the pseudos you provide in `_static/pseudos`:
```
Si_ONCV.upf
H_NCP19_PBE_OTF.usp  
```

**Comment by tomdemeyere** [comment]
@JacekDziedzic There was indeed a tiny mistake in jupyter_t1 but not what you reported, did you forget to switch branch or to pull?

**Comment by JacekDziedzic** [comment]
I used `improved_ase_tutorial`, as that's the one that has seen recent changes. What should I be using?

**Comment by tomdemeyere** [comment]
Yes that's the one, not sure what is happening, if you look at "files changed" in this PR you will see:

image

Which is what is expected

**Comment by JacekDziedzic** [comment]
Hi again @tomdemeyere. I cloned from scratch and checked out `improved_ase_tutorial`. I am now crashing at the original error again -- the launcher script is not being found. Presumably because we are operating from `t1_onetep_silane`, which is now at the same level as `_static`, so it should be looking at `../_static/launch_onetep.sh`. It is, however, looking for `../../_static/launch_onetep.sh`. 

See the attached:

tutorial_1_borked

Can you test that this runs to completion for you?

**Comment by JacekDziedzic** [comment]
@tomdemeyere ?

**Comment by tomdemeyere** [comment]
@JacekDziedzic Sorry for the delay! Indeed, changing to "../" instead of "../../" fixes the problem. It was a typo in t1, but apparently not the others. It runs ok on your machine now

**Comment by JacekDziedzic** [comment]
Thanks @tomdemeyere. T1 now runs for me, with the caveat that bulk silicon is still having NGWF convergence issues. I'll investigate.

**Comment by JacekDziedzic** [comment]
Hi @tomdemeyere . Thanks for your patience. I am somewhere near the end of tutorial 4. I have a few comments.

(1) To get bulk Si to converge, could you please add "k_zero -1.0" to the list of parameters, with a comment "! turn on adaptive kinetic energy preconditioning to improve NGWF convergence"? It's an experimental functionality that adjusts k_zero on the fly, it works wonders on this system.

(2) For sucrose, I get

```
/home/jacek/mambaforge/envs/newconda/lib/python3.12/site-packages/ase/data/pubchem.py:80: UserWarning: The structure "sucrose" has more than one conformer in PubChem. By default, the first conformer is returned, please ensure you are using the structure you intend to or use the `ase.data.pubchem.pubchem_conformer_search` function
  warnings.warn(
```

I understand what that means. Is there anything you can add to the tutorial to have it choose the same conformer explicitly and not produce the warning so that we don't scare users? Maybe a short comment in the notebook that explains that it's doing that?

(3) Your sucrose **doesn't converge**, at least not for me. It's the really bad values for `ngwf_cg_max_step` and `lnv_cg_max_step` (probably the first one) that completely break convergence -- look at the `step` column:

```
|iter|    RMS Gradient   |       Total Energy    |   Step   |   Commutator |         Change        |
   1     0.00322812635597     -251.85647011149644   0.150000  0.00087969000     -251.85647011149644
   2     0.00244523783810     -253.14459947565217   0.150000  0.00023865000       -1.28812936415574
   3     0.00190144809341     -253.89094371200835   0.150000  0.00011253000       -0.74634423635618
   4     0.00151487258241     -254.34735172455100   0.150000  0.00009588000       -0.45640801254265
   5     0.00123280523265     -254.63992555768701   0.150000  0.00008493000       -0.29257383313600
   6     0.00102224854968     -254.83540060234299   0.150000  0.00007460200       -0.19547504465598
   7     0.00086162414802     -254.97085106979586   0.150000  0.00006550300       -0.13545046745287
   8     0.00073648870932     -255.06772744210275   0.150000  0.00005782200       -0.09687637230689
   9     0.00063704008764     -255.13891447193839   0.150000  0.00005138800       -0.07118702983564
  10     0.00055654683505     -255.19243305404365   0.150000  0.00004596300       -0.05351858210526
  11     0.00049032540090     -255.23344859914610   0.150000  0.00004133400       -0.04101554510245
  12     0.00043506648329     -255.26539462120365   0.150000  0.00003734000       -0.03194602205755
  13     0.00038839004926     -255.29062033469376   0.150000  0.00003385900       -0.02522571349010
  14     0.00034855138640     -255.31077511399633   0.150000  0.00003080100       -0.02015477930257
  15     0.00031424603527     -255.32704356062899   0.150000  0.00002809700       -0.01626844663267
  16     0.00028447978550     -255.34029341929448   0.150000  0.00002569500       -0.01324985866549
  17     0.00025848108274     -255.35117127593549   0.150000  0.00002355200       -0.01087785664100
  18     0.00023564115375     -255.36016609651301   0.150000  0.00002163500       -0.00899482057753
  19     0.00021547263125     -255.36765243228490   0.150000  0.00001991600       -0.00748633577189
  20     0.00019758024476     -255.37392044496997   0.150000  0.00001837100       -0.00626801268507
  21     0.00018163981944     -255.37919717208939   0.150000  0.00001698000       -0.00527672711942
  22     0.00016738290497     -255.38366185059604   0.150000  0.00001572600       -0.00446467850665
  23     0.00015458530950     -255.38745712508046   0.150000  0.00001459300       -0.00379527448442
  24     0.00014305841490     -255.39069735497401   0.150000  0.00001356900       -0.00324022989355
  25     0.00013264250555     -255.39347484437945   0.150000  0.00001264200       -0.00277748940545
  26     0.00012320155470     -255.39586456352521   0.150000  0.00001180200       -0.00238971914575
  27     0.00011461910521     -255.39792776028102   0.150000  0.00001104000       -0.00206319675581
  28     0.00010679498348     -255.39971474639785   0.150000  0.00001034900       -0.00178698611683
  29     0.00009964264438     -255.40126706374301   0.150000  0.00000972020       -0.00155231734516
  30     0.00009308701582     -255.40261918049330   0.150000  0.00000914870       -0.00135211675030
```

When I drop these keywords, it converges fine in 11 steps:

```
|iter|    RMS Gradient   |       Total Energy    |   Step   |   Commutator |         Change        |
   1     0.00322342241237     -251.86158819670729   0.595945  0.00014220000     -251.86158819670729
   2     0.00116512678532     -254.81485109572935   0.618328  0.00006031400       -2.95326289902206
   3     0.00066386311426     -255.22777522169127   0.589867  0.00002947200       -0.41292412596192
   4     0.00031159980755     -255.35751355156594   0.729683  0.00002099100       -0.12973832987467
   5     0.00019134919653     -255.39276701986034   0.547122  0.00001178800       -0.03525346829440
   6     0.00012387693251     -255.40219069889849   0.934568  0.00000309130       -0.00942367903815
   7     0.00008996575779     -255.40901609720825   0.646041  0.00000620960       -0.00682539830976
   8     0.00005261058399     -255.41155384666209   0.624033  0.00000500920       -0.00253774945384
   9     0.00002415501592     -255.41241351003967   0.533330  0.00000394270       -0.00085966337758
  10     0.00000289913581     -255.41256294197751   0.026287  0.00000176830       -0.00014943193784
  11     0.00000172330749     -255.41256307579329   0.000000  0.00000014855       -0.00000013381577 <-- CG CONVERGED

```

Why did you change these, and other, settings from the non-ASE tutorial at https://tutorials.onetep.org/T4_geometry_optimization.html? We already provide a .dat file, why not copy the settings from there and just use an ASE-generated structure?

I'll take a look at T10, could you fix these in the meantime?

**Comment by JacekDziedzic** [comment]
Hi @tomdemeyere. Thanks, but this latest "fix" breaks sucrose entirely, see below. Could you **please test these things** before you push them? 
ase_error_3

**Comment by tomdemeyere** [comment]
This is an ASE version related problem, I do test things before pushing

Let me make sure that does not happen to anyone.

**Comment by JacekDziedzic** [comment]
> This is an ASE version related problem

... but I do have the newest version, or at least had one two weeks ago, updated as per your instructions.

> I do test things before pushing

In that case I'm sorry! How can it work for you if we both have the new ASE version?

LMK when you're ready for me to re-test this.

**Comment by JacekDziedzic** [comment]
@tomdemeyere -- please let me know when you're ready for me to re-test this.

**Comment by tomdemeyere** [comment]
@JacekDziedzic 

It seems to work fine on the machine, although I am not completely sure about the sucrose and the calculation in t10, because it takes to much time. The geometry optimisations do progress

**Comment by JacekDziedzic** [comment]
@tomdemeyere -- sucrose doesn't geom-converge at all now, it gives up after the default 30 geom iterations. The original (non-ASE) sucrose tutorials converges nicely in 15 geom iterations. We can't be showing that to our tutees! This happens because you changed two things relative to the original tutorial -- the structure, and the pseudos. I understand changing the structure -- we want to show off the cool capability of just getting the structure from a database. But there is no need to change from ONETEP-default pseudos. Once I put in your structure and the original pseudos, it converges within 20 geom iterations. I think that's fair. 

Could you please replace the ultrasofts you added with the ONETEP-provided `carbon.recpot`, `hydrogen.recpot` and `oxygen.recpot`?

Another worry I have is that the ASE tutorial assumes that it got git-cloned from the repository -- it looks for pseudos and your launcher in `_static`. But I presume most users work with the web-based version on www.onetep.org. They will just download the jupyter notebook file, without the stuff in `_static`. Could you add a brief explanation to each tutorial that further files must be downloaded (pseudos, ASE launcher) and put in a suitably located directory? Preferably with links to these files -- just like the original T4 links to `sucrose.dat`?

I'll move to the last tutorial now.

**Comment by tomdemeyere** [comment]
>> Could you please replace the ultrasofts you added with the ONETEP-provided carbon.recpot, hydrogen.recpot and oxygen.recpot?

During the coding retreat these have been changed due to licensing issues, are we sure about distributing the recpot openly?

**Comment by tomdemeyere** [comment]
From the ONETEP website:

"Older versions of CASTEP code contained a full library of suitable .recpot potentials, but this Library is owned by BIOVIA and we cannot distribute it with the Academic version."

These might not be the same after all, we have to check

**Comment by cksdaddy** [comment]
Dear Tom,
I feel we are very close now to completing the ASE tutorials. It seems that the only missing thing is to have the correct pseudos (the ones that we distribute with ONETEP already (in the pseudo directory)). Could you please use these in the ASE tutorials as they are tested to converge the calculations. We are not allowed to distribute the CASTEP on the fly ones.
With best wishes,
Chris

-------------------------------------------------------------------------
http://www.soton.ac.uk/chemistry/about/staff/cks.page
Chris-Kriton Skylaris FRSC FHEA
Professor of Computational Chemistry
School of Chemistry, University of Southampton,
Highfield, Southampton SO17 1BJ, UK
Tel. +44 (0)2380 599381
--------------------------------------------------------------------------

**Comment by JacekDziedzic** [comment]
@tomdemeyere -- you misunderstand. I'm not asking you to use the CASTEP **on-the-fly** pseudopotentials, neither am I asking you to use the full CASTEP pseudo library. I'm asking to switch to the ones we've been distributing for two decades with the academic version, the ones in `./pseudo`. Like the ones that are _already_ under `./_static/tutorial_6`.

**Comment by JacekDziedzic** [comment]
Ditto for the last tutorial, the ASE-ified version of T10 (stress). Please just use the recpots already provided with it (they are in `_static/tutorial_10`).

**Comment by tomdemeyere** [comment]
Switched to recpot and added a downloadable link in T2 + informations in all tutorials

**Comment by JacekDziedzic** [comment]
Thanks for the rapid response. I will test from scratch tomorrow morning.

**Comment by JacekDziedzic** [comment]
Hi @tomdemeyere. Happy to report it all works for me and converges neatly. Thanks for making these changes and for bearing with me. I appreciate the new explanations, and the fact that keywords are now specified explicitly rather than borrowed from `sucrose.dat` is also a welcome improvement.

One last change I would ask for is to rename the T4 directory to have `t4` in the name, and not `t2` so that we don't confuse people. Once this is in place, I'm happy to merge.

**Comment by tomdemeyere** [comment]
@JacekDziedzic Yes I think that was an attempt of anticipating Nicholas Hine desire to change the numbering, but I think that did not happen after all.

This should be ok!

**Comment by JacekDziedzic** [comment]
Merged! It should appear on the website within a few minutes. We could then discuss and fine-tune the placement of these tutorials. Chris initially suggested something like

Tutorial 1...
Tutorial 1* ASE version
Tutorial 2
Tutorial 3
Tutorial 4
Tutorial 4* ASE version

and so on, but given that there's a separate introductory tutorial for ASE (Tutorial 2) that should be done first by users interested in ASE, I think it would maybe be better to put the ASE-ified tutorials at the bottom, and only indicate earlier that they are available.

**Comment by JacekDziedzic** [comment]
@tomdemeyere -- they deployed, but we can't access them, as they are not listed in the index. Can you suitably edit `index.rst` (I presume) to show them under the list of "usual" tutorials?

**Comment by nickhine** [comment]
Just gave these a go to check I understand the structure. Nice work - easy to use.

A couple of comments:
1) when running eg the convergence tests, the results are all written to the same place ie t1_onetep_silane/onetep.out and the only thing retained is the total energy. Would it not be more instructive to write them each to different filenames so the user could have a look if they were interested? That would more closely match real-world usage I think.

2) I like that we have settings that make bulk silicon converge in 10 iterations, with just 4 NGWFs per atom and with no fiddling with NGWF  initialisation - that was always a problem in the past. However, has anyone checked how "robust" the convergence is for the silicon total energy? Because it could end up being one of the calculations many suspicious-minded people try when testing onetep for the first time, it would be nice to be confident that the total energy matches well with what you get from a plane-wave calculation and that it is not heavily dependent on arbitrary parameters such as kzero or initial NGWFs.

**Comment by nickhine** [comment]
Finally:

3. Are we happy with the fact that perfectly symmetry-obeying bulk silicon (for which forces should be zero) generates a maximum force of 0.018 eV/ang? That seems somewhat large. I presume it is the result of the grid (45^3) not being commensurate with the atoms, so it is a form of eggbox effect. Maybe we should frame it as a question for users to consider?

**Comment by cksdaddy** [comment]
We could put a note saying this is the case and if people want to reduce the force they could increase the NGWF radii and the psinc KE cutoff.

-------------------------------------------------------------------------
http://www.soton.ac.uk/chemistry/about/staff/cks.page
Chris-Kriton Skylaris FRSC FHEA
Professor of Computational Chemistry
School of Chemistry, University of Southampton,
Highfield, Southampton SO17 1BJ, UK
Tel. +44 (0)2380 599381
--------------------------------------------------------------------------

**Comment by nickhine** [comment]
I just noticed by comparison to [https://tutorials.onetep.org/T1_spe_silicon.html](https://tutorials.onetep.org/T1_spe_silicon.html) that actually this is the point of the exercise and the original tutorial recommended people to look at this, realise it is an issue, and make the cell bigger if they want to fix it. However, this also refers to needing to go to 9 NGWFs to get convergence, which seems no longer the case with kzero: -1. Shall we update this text too?

>  An example input file for this job can be found on the tutorial web page. The calculation should take around a few minutes. Feel free to stop it as soon as you see what is happening, since you will find that the calculation fails to converge: the RMS gradient remains stuck above the threshold for convergence. Likewise, the total energy will not converge to a fixed value. Make a copy of your output and modify the NGWF radius in the input file to 8.0 Bohr and the number of NGWFs per Si atom to 9. This introduces NGWFs with d-like symmetry rather than just s and p, allowing much more variational freedom. You should now find the calculation converges nicely, but will take rather longer to run.

**Comment by nickhine** [comment]
Tweaked in commit just now.

**Comment by JacekDziedzic** [comment]
Thanks for the fix, @nickhine. I agree with your point 1 re separate directories. @tomdemeyere, could you make that simple change?

As for point 2, this is how it converged with adaptive K-zero:
```
|iter|    RMS Gradient   |       Total Energy    |   Step   |   Commutator |         Change        |
   1     0.00034136228928     -249.73414923185078   5.107777  0.00000015605     -249.73414923185078
   2     0.00011266049652     -251.47274555082663   6.598575  0.00000112420       -1.73859631897585
   3     0.00006986697228     -251.71631353263047   6.154965  0.00000100710       -0.24356798180384
   4     0.00004172974864     -251.79916865932938   6.066373  0.00000063202       -0.08285512669892
   5     0.00000984206704     -251.82434560901257   1.088570  0.00000096280       -0.02517694968319
   6     0.00000556134855     -251.82460800885258   0.252843  0.00000014309       -0.00026239984001
   7     0.00000443273504     -251.82462344364211   0.092269  0.00000004789       -0.00001543478953
   8     0.00000386814572     -251.82462671913925   0.043920  0.00000002745       -0.00000327549714
   9     0.00000351393704     -251.82462784540184   0.310584  0.00000001771       -0.00000112626259
  10     0.00000238686199     -251.82463546797160   0.050553  0.00000003414       -0.00000762256977
  11     0.00000208117659     -251.82463584320575   0.021204  0.00000001534       -0.00000037523415
  12     0.00000188109150     -251.82463597355218   0.000000  0.00000000996       -0.00000013034642 <-- CG CONVERGED
```
and here is the struggle without it (so, default k zero):
```
   1     0.00056180218854     -249.73414923185015   1.666131  0.00000015605     -249.73414923185015
   2     0.00037853993130     -251.26165329822354   1.191619  0.00000109730       -1.52750406637338
   3     0.00014886524079     -251.75850982426178   1.846805  0.00000071020       -0.49685652603824
   4     0.00009311291798     -251.87755898798713   1.638929  0.00000076857       -0.11904916372535
   5     0.00005598340356     -251.91798307743778   1.221031  0.00000047862       -0.04042408945065
   6     0.00003259312287     -251.92892013320827   2.044963  0.00000017644       -0.01093705577048
   7     0.00003260783357     -251.93436025606317   3.843266  0.00000022990       -0.00544012285491
   8     0.00005219002229     -251.93694107594402   1.544820  0.00000080948       -0.00258081988085
   9     0.00002907259814     -251.94908584543640   1.117439  0.00000027488       -0.01214476949238
  10     0.00000818875203     -251.95176831322976   1.345945  0.00000017866       -0.00268246779336
  11     0.00000896831887     -251.95198845031138   1.086670  0.00000017216       -0.00022013708161
  12     0.00001109557793     -251.95210898992585   2.311416  0.00000024330       -0.00012053961447
  13     0.00001054335788     -251.95288398763967   2.316500  0.00000021274       -0.00077499771382
  14     0.00000814285679     -251.95342256790809   1.820892  0.00000036359       -0.00053858026843
  15     0.00000757295394     -251.95359887485679   2.427347  0.00000038973       -0.00017630694870
  16     0.00000796958002     -251.95363937032511   1.539784  0.00000050906       -0.00004049546831
  17     0.00000789516024     -251.95391710880972   6.554419  0.00000018502       -0.00027773848461
  18     0.00001552204215     -251.95464795597769   1.173027  0.00000074592       -0.00073084716797
  19     0.00000578839404     -251.95546043659652   1.814101  0.00000015091       -0.00081248061883
  20     0.00000793326609     -251.95562223262621   2.521603  0.00000020354       -0.00016179602969
  21     0.00001790783511     -251.95585226582890   1.523193  0.00000071014       -0.00023003320268
  22     0.00000792378630     -251.95725663756335   1.236225  0.00000014979       -0.00140437173445
  23     0.00000584039267     -251.95747043647623   7.467367  0.00000013552       -0.00021379891288
  24     0.00002359171961     -251.95751970410404   1.043486  0.00000105370       -0.00004926762782
  25     0.00000662534635     -251.95919245008781   2.502684  0.00000009220       -0.00167274598377
  26     0.00000927789476     -251.95947175417916   0.001800  0.00000021400       -0.00027930409135
  27     0.00000927944511     -251.95947254818478  -0.002046  0.00000002509       -0.00000079400562
  28     0.00000929240117     -251.95947062685713   1.711206  0.00000001742        0.00000192132765
  29     0.00000702799635     -251.95988161618044   4.046658  0.00000012799       -0.00041098932331
  30     0.00001332345414     -251.96021147029762   1.128324  0.00000047648       -0.00032985411718
  31     0.00000000000000     -251.96078311549613   0.000000  0.00000007760       -0.00057164519851  <-- MAXIT_NGWF_CG EXCEEDED
```

The difference in the total energy is substantial, but wouldn't we be expecting that, given that we've changed the ansatz somewhat, eliminating those pesky high-k vectors in KE? Shouldn't we instead be looking at total energy differences? E.g. silicon vs silicon with defect? I would think these'd be more meaningful.

In any case, @cksdaddy and I were planning to discuss this with @Chengcheng-Xiao during the Masterclass, maybe you could join us, @nickhine?

**Comment by nickhine** [comment]
Tutorial 4 is doing a geometry optimisation with several settings that have no longer have any real impact, due to changes to the defaults:
```
    "penparam": 8,
    "minit_lnv": 4,
    "maxit_lnv": 8,
    "maxit_ngwf_cg": 40,
    "ngwf_cg_max_step": 6.0,
    "lnv_cg_max_step": 5.0, 
    "geom_force_tol": 2e-3,
    "geom_maxiter": 30,
```
Pretty sure all these could be dispensed with, with no noticeable change in the output.

Also, perhaps we could add something after the sucrose optimisation so that people can see what it did? Optimising for an hour just to get a total energy seems a bit anticlimactic! Here's some code that visualises before-and-after geometries:

```
from ase.visualize import view
from ase.io import read
sucrose_opt = read(f"{calc_dir}/onetep.out")
v=view([sucrose,sucrose_opt],viewer='ngl')
v.view.add_ball_and_stick()
v
```

**Comment by JacekDziedzic** [comment]
I'll re-run sucrose without these keywords to see if it still behaves. If so, we could remove them from the original T4 too, I think that's where they came from. @tomdemeyere, could you add @nickhine's visualization snippet?

**Comment by JacekDziedzic** [comment]
I can confirm that removing the keywords @nickhine listed doesn't prevent sucrose from geom-convering. It still takes 20 BFGS steps, and convergence under less than an hour on my machine. @tomdemeyere, could you please remove these keywords from both the original T4 and from the ASE-ified version, please?

**Comment by JacekDziedzic** [comment]
@tomdemeyere -- can I chase you about the below? I'm sure you're busy, but the Masterclass starts tomorrow.

Jacek Dziedzic wrote:
@tomdemeyere -- they deployed, but we can't access them, as they are not listed in the index. Can you suitably edit `index.rst` (I presume) to show them under the list of "usual" tutorials?
